[TSD] Annotates block structures settings and waffle switches

nizarmah opened the pull request:
Documents the following block structure related settings and toggles:
* [`BLOCK_STRUCTURES_SETTINGS`](https://docs.google.com/spreadsheets/d/1xWbEL6oNu6D84WKBs3aViLRM40xk-vmnmmGAeLyR09A/edit?ts=6008a109&pli=1#gid=0&range=A246)
* [`BLOCK_STRUCTURES_SETTINGS['PRUNING_ACTIVE']`](https://docs.google.com/spreadsheets/d/1xWbEL6oNu6D84WKBs3aViLRM40xk-vmnmmGAeLyR09A/edit?ts=6008a109&pli=1#gid=0&range=A4)
* [`BLOCK_STRUCTURES_SETTINGS['COURSE_PUBLISH_TASK_DELAY']`](https://docs.google.com/spreadsheets/d/1xWbEL6oNu6D84WKBs3aViLRM40xk-vmnmmGAeLyR09A/edit?ts=6008a109&pli=1#gid=0&range=A247)
* [`block_structure.invalidate_cache_on_publish`](https://docs.google.com/spreadsheets/d/1xWbEL6oNu6D84WKBs3aViLRM40xk-vmnmmGAeLyR09A/edit?ts=6008a109&pli=1#gid=0&range=A191)
* [`block_structure.storage_backing_for_cache`](https://docs.google.com/spreadsheets/d/1xWbEL6oNu6D84WKBs3aViLRM40xk-vmnmmGAeLyR09A/edit?ts=6008a109&pli=1#gid=0&range=A193)
* [`block_structure.raise_error_when_not_found`](https://docs.google.com/spreadsheets/d/1xWbEL6oNu6D84WKBs3aViLRM40xk-vmnmmGAeLyR09A/edit?ts=6008a109&pli=1#gid=0&range=A192)

_This pull request also updates some other block structure related settings which previously had a good description. However, they were updated for consistency._

edx-webhook commented:
Thanks for the pull request, @nizarmah! I've created [OSPR-5561](https://openedx.atlassian.net/browse/OSPR-5561) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

natabene commented:
@nizarmah Thank you for your contribution. @robrap Please review.

nizarmah commented:
Sorry about that, but moving this back to "Draft" pull request since I decided to include:
* [`BLOCK_STRUCTURES_SETTINGS`](https://docs.google.com/spreadsheets/d/1xWbEL6oNu6D84WKBs3aViLRM40xk-vmnmmGAeLyR09A/edit?ts=6008a109&pli=1#gid=0&range=A246)
* [`BLOCK_STRUCTURES_SETTINGS['PRUNING_ACTIVE']`](https://docs.google.com/spreadsheets/d/1xWbEL6oNu6D84WKBs3aViLRM40xk-vmnmmGAeLyR09A/edit?ts=6008a109&pli=1#gid=0&range=A4)

nizarmah commented on openedx/core/djangoapps/content/block_structure/config/__init__.py:
```diff
@@ -10,12 +10,41 @@
+# .. toggle_description: Raises an error if the requested block structure does not exist in block
+#   structure store, or if it is outdated. Block structure store refers to both cache and storage,
+#   if enabled.
+# .. toggle_use_cases: temporary
```
I know this sounds weird, since this has been there for 4 years, but there's a [comment](https://github.com/edx/edx-platform/pull/14512/files#r103130819) from @nasthagiri, author of the `RAISE_ERROR_WHEN_NOT_FOUND` change mentioning that the switch is temporary in #14512.

> **Note to reviewers**: We now plan to raise an error instead, if the block-structure is not available when requested. This is now an unexpected situation since the block-structure now has permanent storage. We honor the `RAISE_ERROR_WHEN_NOT_FOUND` waffle switch in the interim as we transition over to the new strategy.

nizarmah commented on openedx/core/djangoapps/content/block_structure/config/__init__.py:
```diff
@@ -10,12 +10,41 @@
+# .. toggle_description: Raises an error if the requested block structure does not exist in block
+#   structure store, or if it is outdated. Block structure store refers to both cache and storage,
+#   if enabled.
+# .. toggle_use_cases: temporary
```
I know this sounds weird, since this has been there for 4 years, but there's a [comment](https://github.com/edx/edx-platform/pull/14512/files#r103130819) from @nasthagiri, author of the `RAISE_ERROR_WHEN_NOT_FOUND` change mentioning that the switch is temporary in #14512.

> **Note to reviewers**: We now plan to raise an error instead, if the block-structure is not available when requested. This is now an unexpected situation since the block-structure now has permanent storage. We honor the `RAISE_ERROR_WHEN_NOT_FOUND` waffle switch in the interim as we transition over to the new strategy.

robrap replied:
@nasthagiri: Is this a candidate for DEPR now? Do you know if the _transition_ referred to in the comment above is complete?
@nizarmah: I would add a `.. toggle_warning` with a note that this flag is deprecated and will be removed at some point.  For target date, I am fine with using the 3 month rule from https://edx.readthedocs.io/projects/edx-toggles/en/latest/how_to/documenting_new_feature_toggles.html from the creation date.  This will potentially just bring this to someone's attention as very ready to DEPR and remove.  We'll see if @nasthagiri feels otherwise.
Thank you.

nizarmah replied:
Thanks for the quick reply @robrap! I followed your suggestions with the commit [`b7fa3be`](https://github.com/edx/edx-platform/pull/26175/commits/b7fa3bec84db96278f8fb50d6926d12b8d648487).

I'll keep my eye out for a reply from @nasthagiri, and apply the necessary changes 👍🏼

robrap replied:
Great.
1. I think it is toggle_warning (singular).
2. Consider adding a link to the comment you posted from Nimisha for more details in the warning.
2. If we don’t hear back tomorrow, just mark this resolved.
Thanks.

nizarmah replied:
> I think it is toggle_warning (singular).

I folowed [`feature_toggle_annotations.yaml`](https://github.com/edx/code-annotations/blob/70d076aa843ce5ca688e08623e6d40bc6d4aa241/code_annotations/contrib/config/feature_toggle_annotations.yaml#L20).

> Consider adding a link to the comment you posted from Nimisha for more details in the warning.

Yeah, that's a great idea 👍🏼 

> If we don’t hear back tomorrow, just mark this resolved.

Alright then 😃

robrap replied:
Yikes. Thanks.
@regisb: For warnings, we are using singular for settings and plural for toggles. See https://github.com/edx/code-annotations/blob/70d076aa843ce5ca688e08623e6d40bc6d4aa241/code_annotations/contrib/config/setting_annotations.yaml
I’d like this to be consistent. I probably prefer singular, but we probably have many more toggles, so we should probably go for simplest. Do you mind updating? Do you think we should add for ticket (or tickets?) optionally in settings?

nasthagiri replied:
@nizarmah Thank you for raising the question regarding the `RAISE_ERROR_WHEN_NOT_FOUND` toggle. You rightly discovered that it was created to be used only for a short-term rollout period, with the intent to remove it.

So yes, this toggle should be DEPR-ed and removed. Looking in our production environment, though, I am a bit confused why it is not enabled. Perhaps there was a problem when it was turned on and so we kept it off.

At this point, given the age of this block_structure code, we'll want to just remove the toggle and the code (since it's been disabled this entire time).

regisb replied:
>  I probably prefer singular, but we probably have many more toggles, so we should probably go for simplest. Do you mind updating?

Yes, I prefer singular, too. But I'm not so keen on suddenly renaming the field, which would cause many linting failures. The right deprecation process would be the same as for databases:

1. Add a `.. toggle_warning:` field which would be handled the same way as `.. toggle_warnings:`, including in the Sphinx extension.
2. Rename all `.. toggle_warnings:` to `.. toggle_warning:`.
3. Drop the `.. toggle_warnings:` field.

nizarmah replied:
@nasthagiri thanks for your reply! That sounds good 👍🏼 
I've went with @robrap's suggestion to indicate its deprecation in the warning annotation.

Let me know, please, if you'd like me to make any further changes though 😄

nizarmah commented on openedx/core/djangoapps/content/block_structure/config/__init__.py:
```diff
@@ -10,12 +10,41 @@
+#   if enabled.
+# .. toggle_use_cases: temporary
+# .. toggle_creation_date: 2017-02-23
+# .. toggle_target_removal_date: 2022-01-01
```
Since this is a temporary switch, as [mentioned earlier](https://github.com/edx/edx-platform/pull/26175/files#r564816707), I had to set a removal date. However, I'm in no seat to choose that removal date.

So, @robrap do you have any suggestions on what to do in this case?

robrap replied:
Answered in https://github.com/edx/edx-platform/pull/26175#discussion_r564842861

robrap commented on openedx/core/djangoapps/content/block_structure/config/__init__.py:
```diff
@@ -10,12 +10,42 @@
 from .models import BlockStructureConfiguration
 
 # Switches
+# .. toggle_name: INVALIDATE_CACHE_ON_PUBLISH
```
Should be `block_structure.invalidate_cache_on_publish`.

robrap commented on openedx/core/djangoapps/content/block_structure/config/__init__.py:
```diff
@@ -10,12 +10,42 @@
+# .. toggle_name: INVALIDATE_CACHE_ON_PUBLISH
+# .. toggle_implementation: WaffleSwitch
+# .. toggle_default: False
+# .. toggle_description: Invalidates block structure cache when changes to courses are published.
```
I wonder if we could provide more help as to when I would want to have this off or on?  Was this meant to be temporary, or is this required as an option for different use cases?

robrap replied:
Thanks. Very helpful.
1. Could you begin with “When enabled...” or something like that?
2. Also, maybe replacing “This toggle comes in handy...” with something like “Enabling this toggle is useful for...”
3. Maybe replace “It is also very useful when...” with “It is also useful to enable this when...”.

nizarmah replied:
I really appreciate your patience with those reviews by the way. Sorry about that, but I'm not the best with writing documentation, and your comments are exactly what I needed 😃

robrap replied:
You are doing a great job. Thanks.

nizarmah commented:
jenkins run python

robrap commented on openedx/core/djangoapps/content/block_structure/config/__init__.py:
```diff
@@ -10,12 +10,42 @@
+# .. toggle_name: STORAGE_BACKING_FOR_CACHE
+# .. toggle_implementation: WaffleSwitch
+# .. toggle_default: False
+# .. toggle_description: Enables retrieving block structures from storage instead of regenerating
```
What about something like?:
```
When enabled, stores block structures in a more permanent storage like a database, that provides an additional backup for cache misses, instead of regenerating...
```

robrap commented on openedx/core/djangoapps/content/block_structure/config/__init__.py:
```diff
@@ -10,12 +10,42 @@
+# .. toggle_default: False
+# .. toggle_description: Enables retrieving block structures from storage instead of regenerating
+#   block structures when they aren't available in cache.
+# .. toggle_warnings: Depends on the storage class, `STORAGE_CLASS`, and storage keyword arguments,
```
If you use something like the following, they could potentially search the docs more easily for details.
```
Depends on BLOCK_STRUCTURES_SETTINGS['STORAGE_CLASS'] and ...
```

nizarmah replied:
Yeah, I don't know what I was thinking to be honest 😅 Addressed this 👍🏼

robrap reviewed (commented):
Thanks for all the docs!

nizarmah commented on lms/envs/common.py:
```diff
@@ -2347,21 +2347,54 @@ def _make_locale_paths(settings):  # pylint: disable=missing-function-docstring
+#   }
+# .. setting_description: Stores all the settings used by block structures and block structure
+#   related tasks, such as updating the cached block structures.
+# .. setting_warnings: This setting is also used to specify the `STORAGE_CLASS` and
```
https://github.com/edx/edx-platform/blob/ad5aa6cd7a6d61ca64521c92595fef85c5ba464c/openedx/core/djangoapps/content/block_structure/models.py#L62-L69

@robrap unfortunately, `BLOCK_STRUCTURES_SETTINGS['STORAGE_CLASS']` and `BLOCK_STRUCTURES_SETTINGS['STORAGE_KWARGS']` are directly accessed and used 😔 

I can move that information though, there?
Or maybe, throw an `ImproperlyConfigured` exception?

robrap replied:
I think we should either add the annotations where the setting is retrieved, or add new defaults that match the retrieval code and annotate. It may make sense to keep these annotations together? Thoughts?

nizarmah replied:
I do prefer adding new defaults and annotate 👍🏼 

However, this will need testing, to verify it. So, so I'll need to spin up two sandboxes to test two cases.
The first being with an S3 bucket, and the second being the file system.

It should be safe to specify something like, with the necessary annotations
```
STORAGE_CLASS = "django.core.files.storage.FileSystemStorage"
STORAGE_KWARGS = {}
DIRECTORY_PREFIX = "/tmp/courses/"
```

Should I move forward with this change?

robrap replied:
1. Maybe we should add the annotations where they are being retrieved for now to minimize risk?
2. If you want to follow up with a PR to move the annotations and set defaults, I’ll leave that to you.
3. Is the default for STORAGE_CLASS currently None? Is that a useless default that you are proposing to change, or did I miss this default somewhere?
3. Is DIRECTORY_PREFIX another one? Same applies around minimizing risk and just annotating somewhere that makes sense.

Let me know if this makes sense. Thanks.

nizarmah replied:
> Maybe we should add the annotations where they are being retrieved for now to minimize risk?

Alright, sure :+1: Sounds good. This change should be addressed now 😄 

> Is the default for STORAGE_CLASS currently None? Is that a useless default that you are proposing to change, or did I miss this default somewhere?

Yeah, the default for `STORAGE_CLASS` is currently `None`. And yeah, it is a useless default which I'm proposing to change. Let me know if you would prefer that I change the default value for it, though 👍🏼 

> Is DIRECTORY_PREFIX another one? Same applies around minimizing risk and just annotating somewhere that makes sense.

Yes, `DIRECTORY_PREFIX` is another one, which I forgot about 😆 
I'll annotate it similarly, as you requested 👍🏼

nizarmah reviewed (commented):
@robrap I addressed the changes you requested. Thanks a lot for your thorough review! 😄 

I kept the issues unresolved for you to resolve them once you're satisfied with the changes made 👍🏼

robrap commented on openedx/core/djangoapps/content/block_structure/models.py:
```diff
@@ -63,10 +71,21 @@ def _bs_model_storage():
-    )
+    # .. setting_name: BLOCK_STRUCTURES_SETTINGS['STORAGE_CLASS']
+    # .. setting_default: None
+    # .. setting_description: Specifies the storage used for storage-backed block structure cache.
```
Consider adding your proposed class as an example, unless you think the options are obvious.

nizarmah replied:
The options are somewhat obvious, but I'll add a link to an example instead 👍🏼 This [forum post](https://discuss.openedx.org/t/help-please-very-slow-load-time-10-seconds-for-courseware-on-sections-with-several-subsections-and-xblocks/2998/16) has a wonderful example for the three settings 🙂

robrap commented:
@nizarmah: Thanks for all the great docs.
1. I’ll come back to this tomorrow.
2. Please update the PR description with any new annotations. I’ll need to remember to add to the sign-up sheet tomorrow.
Thank you!

nizarmah reviewed (commented):
@robrap I really appreciate all the effort you're putting into refining the doc changes! 

I addressed the second batch of comments you had. The comments below reflect the changes that were done in the latest commit, [`000a554`](https://github.com/edx/edx-platform/pull/26175/commits/000a55403d5a8150df99861e34767edf959184ea) 😃

farhaanbukhsh commented:
@nizarmah @robrap  I have one suggestion that I missed as a developer, if there was a way of pointing out that I can use `STORAGE_CLASS` and `STORAGE_KWARGS`  in `lms/envs/common.py`. That would save me a lot of time in diving in the code to find out what we can do. :) 

I don't know if I am missing something in the way I approach a setting.

Also is there a way we can document why we are using this setting and what effect it can have?

I mean the docs answer the question of `how` to set these values but doesn't answer the question `why` do I need to do it.

I hope I am able to convey myself. Thanks a lot for all the work here. :smile:

nizarmah commented:
@farhaanbukhsh Hello! 😃 

> I have one suggestion that I missed as a developer, if there was a way of pointing out that I can use `STORAGE_CLASS` and `STORAGE_KWARGS` in lms/envs/common.py. That would save me a lot of time in diving in the code to find out what we can do. :)

@farhaanbukhsh yeah you're right, including the `STORAGE_CLASS` and `STORAGE_KWARGS` in the configuration is an important note, but I think it is implied by the `BLOCK_STRUCTURES_SETTINGS['STORAGE_CLASS']` and `BLOCK_STRUCTURES_SETTINGS['STORAGE_KWARGS']` mentioned in the annotation. 

https://github.com/edx/edx-platform/blob/000a55403d5a8150df99861e34767edf959184ea/openedx/core/djangoapps/content/block_structure/models.py#L77-L95

> I don't know if I am missing something in the way I approach a setting.

Mainly, you don't want to apply the changes to `lms/envs/common.py`, but to the "extra configuration file" that would be used to override the defaults. 

> Also is there a way we can document why we are using this setting and what effect it can have?

I included a link to your forum post, kudos on for that awesome post. But I think you can contribute something to [edx/edx-developer-docs](https://github.com/edx/edx-developer-docs) or [edx/edx-documentation](https://github.com/edx/edx-documentation)

> I mean the docs answer the question of _how_ to set these values but doesn't answer the question of _why_ do I need to do it.

That's a good point, but imho this remains an annotation. So, if we want to have more information added about it, we'll need to create a documentation, and then update the annotations to link to it, instead of linking to the forum post.

---

Hope that answers your questions 👍🏼 The concerns you raised are valid, and I hope I didn't make you feel like I'm dismissing them, but I think those comments would be better addressed in a change to the documentation, directly, instead of the annotations.

The storage-backed block structures solution is something that, in my personal opinion, every instance in production should have. So, having a whole documentation for it might be a better way to express it's necessity and it's use case.

robrap reviewed (commented):
Thanks again.  Minor feedback, and then I think this is ready.

robrap commented:
@farhaanbukhsh and @nizarmah: I wanted to add some thoughts.

> > I have one suggestion that I missed as a developer, if there was a way of pointing out that I can use `STORAGE_CLASS` and `STORAGE_KWARGS` in lms/envs/common.py. That would save me a lot of time in diving in the code to find out what we can do. :)
> 

@farhaanbukhsh: Once this lands, the annotations will all be published in this doc: https://edx.readthedocs.io/projects/edx-platform-technical/en/latest/.  Would this serve the purpose of not having to dive into code?

> > I don't know if I am missing something in the way I approach a setting.

@farhaanbukhsh: I'm not clear if you are asking about general docs on how to use settings, or if this was just an intro to specific questions you had about these settings in particular.

> > Also is there a way we can document why we are using this setting and what effect it can have?
> > I mean the docs answer the question of _how_ to set these values but doesn't answer the question of _why_ do I need to do it.

@farhaanbukhsh: As noted, hopefully the link to your post provides a little more context. If you had something brief and specific in mind for enhancing any of the annotations, why don't you create a PR once this lands with any suggested updates.

> The storage-backed block structures solution is something that, in my personal opinion, every instance in production should have. So, having a whole documentation for it might be a better way to express it's necessity and it's use case.

@nizarmah: Do you think the switch annotation should have a note about this being recommended for Production?

nizarmah commented on openedx/core/djangoapps/content/block_structure/config/__init__.py:
```diff
@@ -10,12 +10,47 @@
+# .. toggle_implementation: WaffleSwitch
+# .. toggle_default: False
+# .. toggle_description: When enabled, stores block structures in a more permanent storage, like
+#   a database, which provides an additional backup for cache misses, instead of regenerating...
```
> The ... at the end was meant to reference the rest of your original sentence. Can you add back the lost text?

I don't know how I missed changing that. Trust me, I'm just as surprised that the `...` is in the text 🤦🏼 

> Could you add the following here, and remove it from the other annotations? If you feel strongly about leaving this text in all the annotations, please update them with this text as well.

Yes, definitely 👍🏼 

I removed it from the other locations, and only kept it for `block_structure.storage_backing_for_cache` because they all depend on it being enabled.

---

Both should be addressed now 👍🏼

nizarmah commented on lms/envs/common.py:
```diff
@@ -2347,21 +2347,44 @@ def _make_locale_paths(settings):  # pylint: disable=missing-function-docstring
+    # .. toggle_default: False
+    # .. toggle_description: When `True`, only a specified number of versions of block structure
+    #   files are kept for each structure, and the rest are cleaned up. The number of versions that
+    #   are kept can be specified in the `BlockStructureConfiguration`, which can be editted in
```
Should be addressed now 👍🏼  🤦🏼

nizarmah reviewed (commented):
> Do you think the switch annotation should have a note about this being recommended for Production?

@robrap Yeah, I added something about that in the latest changes 👍🏼

robrap commented on openedx/core/djangoapps/content/block_structure/config/__init__.py:
```diff
@@ -28,8 +28,12 @@
-#   a database, which provides an additional backup for cache misses, instead of regenerating...
+# .. toggle_description: When enabled, block structures are stored in a more permanent storage,
+#   like a database, which provides an additional backup for cache misses, instead having them
+#   regenerated. The regenration of block structures is a time consuming processs. Therefore,
```
processs => process

nizarmah replied:
I'm really sorry you're having to post such silly comments for my changes 😂 

My pull requests are usually much better, but the last couple of days have been hectic 😔 

Anyway, addressed 👍🏼

robrap reviewed (commented):
Just about there. 😄

robrap commented on openedx/core/djangoapps/content/block_structure/config/__init__.py:
```diff
@@ -10,12 +10,51 @@
+#   regenerated. The regenration of block structures is a time consuming processs. Therefore,
+#   enabling this switch is recommended for Production.
+#   For an example with additional context, see the following forum post:
+#   https://discuss.openedx.org/t/help-please-very-slow-load-time-10-seconds-for-courseware-on-sections-with-several-subsections-and-xblocks/2998/16
```
Sorry.  The quality error, `36: C0301: (line-too-long), : Line too long (148/120)`, is my fault.  
This stinks. Not ideal, but maybe add a pylint disable pragma at the end of the line for now. 😞 

@regisb: Would it make sense to strip pylint disable pragmas, or maybe specific ones like this one, from the annotation text?

nizarmah replied:
Yeah, expected that error to be honest. I should have added the disable line too long lint.

True, it isn't an ideal solution, so hopefully @regisb will be able to help 😄 

For the time being, addressed 👍🏼

regisb replied:
I think we should not be adding links to discussion forums in the source code. Instead, these links should go to the git commit messages. Otherwise we would be adding tons of links (to Slack, Jira, forums) everywhere.

nizarmah replied:
@regisb I addressed your comment and moved the link from the annotation to the commit description 👍🏼 cc @robrap 

Please note that I squashed and force pushed so that it can be merged.

robrap reviewed (approved):
Thanks for all the updates.  If you want to provide a final commit comment in a gihub comment, I can squash and merge once green.  If you want to squash, go for it.

nizarmah commented:
@robrap squashed and provided a commit message 👍🏼 

Thank you so much for being patient with me on those changes! And I apologize for what my lack of sleep resulted in 😂 

Have a wonderful week 😃

robrap commented on openedx/core/djangoapps/content/block_structure/config/__init__.py:
```diff
@@ -10,12 +10,51 @@
+#   regenerated. The regenration of block structures is a time consuming process. Therefore,
+#   enabling this switch is recommended for Production.
+#   For an example with additional context, see the following forum post:
+#   https://discuss.openedx.org/t/help-please-very-slow-load-time-10-seconds-for-courseware-on-sections-with-several-subsections-and-xblocks/2998/16  pylint: disable=line-too-long
```
Ugh. Now it is `36: I0021: (useless-suppression): Useless suppression of 'line-too-long'`. Hopefully we don’t have too linters playing tug-of-war. Make note of what script is failing, in case a different fails if you add useless-suppression. I fear there may not be a combo that makes it happy, but you’ll see. Good luck. If it doesn’t work out, we can do the multi-line url and I’ll let others know about the problem.

Thank you!

nizarmah replied:
Apparently, adding `useless-suppression` to that pragma did not cause additional issues \o/ 🥳 

I'm going to squash, again 😃

robrap replied:
That’s good. But why did we need it at all? Frustrating, but at least it works.
FYI: @jmbowman: Odd issue with pylint violations that failed on line too long, and then complained when we disabled it.

nizarmah replied:
@robrap I think it considered the whole line a pragma, possibly. I did a quick search about any similar issues, I didn't find anything, but there's a chance I might if I spend more time on it.

But from a glance, it seems like when `pylint: ` was added, it might have considered the whole comment as a pragma. I believe they are not considered when the line length is checked, which is what happened here.

robrap commented:
Thank you again. I’ll merge/deploy in the morning.

nasthagiri commented on openedx/core/djangoapps/content/block_structure/config/__init__.py:
```diff
@@ -10,12 +10,51 @@
+#   the latest changes.
+# .. toggle_use_cases: opt_in
+# .. toggle_creation_date: 2017-02-23
+# .. toggle_tickets: https://github.com/edx/edx-platform/pull/14571
```
@nizarmah Not a blocker for this PR, but:

In order to further reduce the complexity in the `block_structure` code. I believe we can DEPR this particular toggle as well.  This toggle also remains disabled in our own production environment.  It was created as a toggle while the team was determining how best to stabilize the `block_structure` framework in production.

The default is to NOT invalidate the cache immediately. That is, to rely on eventual consistency. Some clients may still continue to use the old cached value, while the new cached value is being computed.  This is a more robust solution for a distributed environment. I would be surprised if another Open edX instance really needed this toggle enabled. We can check with the community through a DEPR ticket.

nizarmah replied:
@nasthagiri 

> The default is to NOT invalidate the cache immediately. That is, to rely on eventual consistency. Some clients may still continue to use the old cached value, while the new cached value is being computed. This is a more robust solution for a distributed environment.

I still have a lot to learn, so I'm glad you clarified that 👍🏼 

> I would be surprised if another Open edX instance really needed this toggle enabled. We can check with the community through a DEPR ticket.

I noticed it was enabled on one of our instances at OpenCraft with a claim that it helps guarantee that the storage backed block structures are updated. But after checking your comment and the related code, I went ahead and tested the changes, while storage backed block structures are enabled.

So what this does, **while storage backed block structures are enabled**, is: deletes ALL existing storage backed block structures from BlockStructureStore (storage and/or cache), and then saves the new structure.

But we already have `BLOCK_STRUCTURES_SETTINGS['PRUNING_ACTIVE']`, so I think it should be safe to deprecate that flag. But for now, I won't indicate that in my annotation, if you don't mind.

I ended up removing the lines related to why this toggle should be enabled. I substituted them with information about what happened when storage backed block structures are enabled as well.

```diff
- #   courses are published. Enabling this toggle is useful for non-expiring block structure cache.
- #   It is also useful to enable this when `block_structure.storage_backing_for_cache` is enabled
- #   because it guarantees the storage, containing the storage-backed structures, is updated with
- #   the latest changes.
+ #   courses are published. If `block_structure.storage_backing_for_cache` is active, all block
+ #   structures related to the published course are also cleared from storage.
```

Let me know please if you'd like me to make further changes 👍🏼 cc @robrap

nasthagiri replied:
Sounds good.

nasthagiri commented on openedx/core/djangoapps/content/block_structure/config/__init__.py:
```diff
@@ -10,12 +10,51 @@
+# .. toggle_use_cases: opt_in
+# .. toggle_creation_date: 2017-02-23
+# .. toggle_tickets: https://github.com/edx/edx-platform/pull/14512
 STORAGE_BACKING_FOR_CACHE = WaffleSwitch(
```
@nizarmah Not a blocker for this PR, but:

Similarly to the other toggle above, in order to further reduce the complexity, what do you think about DEPRing this `storage_backing_for_cache` toggle as well?  Is there any reason one would NOT want to have a storage backing? This toggle works on local environment as well as prod, right? Once again, we can announce the removal of this toggle through the DEPR process to double-check.  But there's great value in simply removing these temporary toggles in order to simplify the code.

nizarmah replied:
Yeah, I think that it would be a wonderful idea to enable this in the future, by default.

But it would also be helpful to enable `BLOCK_STRUCTURES_SETTINGS['PRUNING_ACTIVE']` as well, just to guarantee that the storage isn't swarmed by block structures 😂 🤷🏼

regisb commented on lms/envs/common.py:
```diff
@@ -2347,21 +2347,54 @@ def _make_locale_paths(settings):  # pylint: disable=missing-function-docstring
 
-    # Backend storage options
+    # .. toggle_name: BLOCK_STRUCTURES_SETTINGS['PRUNING_ACTIVE']
+    # .. toggle_implementation: SettingDictToggle
```
Yes, I agree that DjangoSetting are more approriate when not using Setting(Dict)Toggle classes.

nasthagiri commented:
@nizarmah Thanks for taking the time to annotate the toggles in this area of the code. This PR now reminds me how many of these toggles were introduced as temporary flags to enable rollout at edX.org and in community instances. This is a testament to this toggle annotation effort for surfacing the temporary nature of these toggles and the path to deprecating them.

(Outside the scope of this particular PR) Would you be interested in helping us follow up with creating DEPR tickets and removing all the binary temporary toggles at this point?  It looks like the following toggles are all temporary and can be now removed:

- BLOCK_STRUCTURES_SETTINGS['PRUNING_ACTIVE']
- block_structure.invalidate_cache_on_publish
- block_structure.storage_backing_for_cache
- block_structure.raise_error_when_not_found

nizarmah commented:
@nasthagiri yeah I'd love to help out and create those DEPR tickets. I'll need a bit of time to do that, however. 

I most likely will need to do that in the next week 👍🏼 

In the meantime, would you be able to, please, provide a quick overview about why `BLOCK_STRUCTURES_SETTINGS['PRUNING_ACTIVE']` is not necessary? Is there an intention to enable it by default?

farhaanbukhsh commented:
> @farhaanbukhsh Hello! 

Hey :smile: 

> 
> > I have one suggestion that I missed as a developer, if there was a way of pointing out that I can use `STORAGE_CLASS` and `STORAGE_KWARGS` in lms/envs/common.py. That would save me a lot of time in diving in the code to find out what we can do. :)
> 
> @farhaanbukhsh yeah you're right, including the `STORAGE_CLASS` and `STORAGE_KWARGS` in the configuration is an important note, but I think it is implied by the `BLOCK_STRUCTURES_SETTINGS['STORAGE_CLASS']` and `BLOCK_STRUCTURES_SETTINGS['STORAGE_KWARGS']` mentioned in the annotation.

That make sense :+1: 

> 
> https://github.com/edx/edx-platform/blob/000a55403d5a8150df99861e34767edf959184ea/openedx/core/djangoapps/content/block_structure/models.py#L77-L95
> 
> > I don't know if I am missing something in the way I approach a setting.
> 
> Mainly, you don't want to apply the changes to `lms/envs/common.py`, but to the "extra configuration file" that would be used to override the defaults.
> 
> > Also is there a way we can document why we are using this setting and what effect it can have?
> 
> I included a link to your forum post, kudos on for that awesome post. But I think you can contribute something to [edx/edx-developer-docs](https://github.com/edx/edx-developer-docs) or [edx/edx-documentation](https://github.com/edx/edx-documentation)
> 

Thanks a lot for pointing it out to me.

> > I mean the docs answer the question of _how_ to set these values but doesn't answer the question of _why_ do I need to do it.
> 
> That's a good point, but imho this remains an annotation. So, if we want to have more information added about it, we'll need to create a documentation, and then update the annotations to link to it, instead of linking to the forum post.
> 
> Hope that answers your questions 👍🏼 The concerns you raised are valid, and I hope I didn't make you feel like I'm dismissing them, but I think those comments would be better addressed in a change to the documentation, directly, instead of the annotations.
> 

Thank you for the patience and clarifying things for me. Makes a lot of sense now.

> The storage-backed block structures solution is something that, in my personal opinion, every instance in production should have. So, having a whole documentation for it might be a better way to express it's necessity and it's use case.

farhaanbukhsh commented:
> @farhaanbukhsh and @nizarmah: I wanted to add some thoughts.
> 
> > > I have one suggestion that I missed as a developer, if there was a way of pointing out that I can use `STORAGE_CLASS` and `STORAGE_KWARGS` in lms/envs/common.py. That would save me a lot of time in diving in the code to find out what we can do. :)
> 
> @farhaanbukhsh: Once this lands, the annotations will all be published in this doc: https://edx.readthedocs.io/projects/edx-platform-technical/en/latest/. Would this serve the purpose of not having to dive into code?
> 

Definately :+1:  I had so much trouble finding anything documented about it :) so kudos to this. :partying_face: 

> > > I don't know if I am missing something in the way I approach a setting.
> 
> @farhaanbukhsh: I'm not clear if you are asking about general docs on how to use settings, or if this was just an intro to specific questions you had about these settings in particular.
> 

I am asking for general doc of how to use it and why should we set this in the first place. But this could be my bias since I really like the optimization and I somehow feel it deserves a blog post of its own.

> > > Also is there a way we can document why we are using this setting and what effect it can have?
> > > I mean the docs answer the question of _how_ to set these values but doesn't answer the question of _why_ do I need to do it.
> 
> @farhaanbukhsh: As noted, hopefully the link to your post provides a little more context. If you had something brief and specific in mind for enhancing any of the annotations, why don't you create a PR once this lands with any suggested updates.
> 

I will have be on the look out of how we can improve this :) so definitely I will make PR. Thanks for the support :heart: 

> > The storage-backed block structures solution is something that, in my personal opinion, every instance in production should have. So, having a whole documentation for it might be a better way to express it's necessity and it's use case.
> 
> @nizarmah: Do you think the switch annotation should have a note about this being recommended for Production?

nasthagiri commented:
Next week for DEPR sounds great!

> In the meantime, would you be able to, please, provide a quick overview about why BLOCK_STRUCTURES_SETTINGS['PRUNING_ACTIVE'] is not necessary? Is there an intention to enable it by default?

When we go through the DEPR process, we can find community input on whether anyone needs us to continue keeping this toggle.

I included it in the list of DEPR based on an [earlier comment](https://github.com/edx/edx-platform/pull/26175#discussion_r565909563) from your investigation:

```
it would also be helpful to enable BLOCK_STRUCTURES_SETTINGS['PRUNING_ACTIVE'] as well, just to guarantee that the storage isn't swarmed by block structures
```

robrap commented on lms/envs/common.py:
```diff
@@ -2347,21 +2347,44 @@ def _make_locale_paths(settings):  # pylint: disable=missing-function-docstring
+    #   files are kept for each structure, and the rest are cleaned up. The number of versions that
+    #   are kept can be specified in the `BlockStructureConfiguration`, which can be edited in
+    #   Django Admin. The default number of versions that are kept is `5`.
+    # .. toggle_use_cases: opt_in
```
If we may DEPR, should we change this to temporary and add a note recommending that everyone use True, which will be the new default (if I understand correctly)?
Or, will we leave all updates to this to when we have a DEPR ticket?

An interesting related question is where to move the docs for the number of versions in admin if this setting is deleted.  You might want to note this in the DEPR ticket so the docs don't get lost in a simple delete.

nizarmah replied:
I think we might as well specify it as temporary, alongside all the toggles that may get deprecated. And if something changes, then we can possibly update the annotation, and include better information on the use cases that would have been mentioned. Do you agree with that, or no?

> An interesting related question is where to move the docs for the number of versions in admin if this setting is deleted.

Yeah this is going to be tough, since having it as an "option" in Django Admin is going to be tough to annotate.

I think this would require documenting it in the docs, do you agree with that, or do you have a different idea? I'd love to hear your thoughts on that 😃 

> You might want to note this in the DEPR ticket so the docs don't get lost in a simple delete.

Will do 👍🏼

robrap commented:
@nizarmah: I think it makes sense to land this sooner than later, and we can update post-DEPR ticket creation. That said, if the proposal of the DEPR ticket is to switch some of the defaults when removing, then it may make sense to add to the description (or warning) that we recommend setting to `True`, rather than leaving the default. Thoughts?

nizarmah commented:
@robrap I have the change ready if you'd like me to push it. I was just finalizing it. I don't have any objection for making the changes post-DEPR ticket creation or pre-DEPR ticket creation.

I would prefer you have the final say here, as you're the more qualified person to do so.

I'm keeping a close eye at the pull request for the next 3 hours. Whatever your decision is, the pull request is either ready or a single push away. 😃

robrap commented:
@nizarmah: I'm not clear on what "the change" means?  Would you mind pushing it as a separate commit, rather than squashing?  Then we can do a quick review and decide from there.

nizarmah commented:
@robrap the change I made was to indicate the deprecation of the flags mentioned [earlier](https://github.com/edx/edx-platform/pull/26175#issuecomment-769080286).

I pushed the change; however, I'm not sure whether we should link to the comment or not, since it was suggested earlier that we include links only in the commit message.

Anyway, let me know if you'd like me to revert the commit 👍🏼

robrap commented on lms/envs/common.py:
```diff
@@ -2347,21 +2347,48 @@ def _make_locale_paths(settings):  # pylint: disable=missing-function-docstring
+    #   files are kept for each structure, and the rest are cleaned up. The number of versions that
+    #   are kept can be specified in the `BlockStructureConfiguration`, which can be edited in
+    #   Django Admin. The default number of versions that are kept is `5`.
+    # .. toggle_warnings: This toggle is deprecated and will be removed at some point.
+    #   It is recommended to set this to `True`, which will be the new default after deprecation.
+    #   For more information, check https://github.com/edx/edx-platform/pull/26175#issuecomment-769080286
```
It seem like we are mostly confident about the recommendation, but unsure of the removal until we get through the DEPR process.  What about something like?:
```suggestion
    # .. toggle_warnings: This toggle will likely be deprecated and removed.
    #   It is recommended to set this to `True`, which will be the new default after deprecation.
    #   The annotation will be updated with the DEPR ticket once that process has started.
```
I don't think you need to link to the issuecomment.  Later, the DEPR ticket should contain all the relevant info.

Thoughts?

nizarmah replied:
Yeah that sounds great 👍🏼 I'll make the changes you recommended. I also noticed that I "copy-pasta"-ed without fixing the values for the other flags 😳 so I'll fix those two and push. Would you like me to squash as well?

robrap commented on openedx/core/djangoapps/content/block_structure/config/__init__.py:
```diff
@@ -16,9 +16,9 @@
-#   It is recommended to set this to `True`, which will be the new default after deprecation.
-#   For more information, check https://github.com/edx/edx-platform/pull/26175#issuecomment-769080286
+# .. toggle_warnings: This switch will likely be deprecated and removed.
+#   It is recommended to keep this as `False`, which will be the new default after deprecation.
```
Yikes.  It looks like most of these switch to True except this one?  Would you mind replying with the appropriate PR comments if you want me to double-check the recommendations?

nizarmah replied:
Definitely 😄 This should be all:
- `block_structure.raise_error_when_not_found`: [`True`](https://github.com/edx/edx-platform/pull/26175#discussion_r565796224)
- `block_structure.invalidate_cache_on_publish`: [`True`](https://github.com/edx/edx-platform/pull/26175#discussion_r565800087)
- `block_structure.storage_backing_for_cache`: [`False`](https://github.com/edx/edx-platform/pull/26175#discussion_r565801078)
- `BLOCK_STRUCTURES_SETTINGS['PRUNING_ACTIVE']` : [`True`](https://github.com/edx/edx-platform/pull/26175#issuecomment-769211933)

robrap replied:
Thanks for those links. That helped a lot. The only one that I read differently is Nimisha's comment about `raise_error_when_not_found`. It sounded to me like she was voting for removing it and leaving it off.

If you aren't sure which way to go, we could drop the recommendation like and ask the question in the DEPR ticket about which way we should go.

nizarmah replied:
@robrap I'm glad the links helped 😄 

Regarding `raise_error_when_not_found`, I think it would be better to refer to this https://github.com/edx/edx-platform/pull/14512/files#r103130819 instead. This comment, on the original pull request, states that they want errors raised. cc @nasthagiri 

It's a little bit confusing, so I guess we might as well drop the recommendation line, what are your thoughts on that?

robrap replied:
Thanks @nizarmah.  I'm going to merge this.  You can discuss and update recommendations when you update to point to the DEPR ticket.

I think that @nasthagiri was saying that since `raise_error_when_not_found` was never turned on for us, it may have either had an issue, or it was forgotten about and hasn't been an issue. We want to remove as many flags as possible.  Removing this as True would mean we'd first need to prove its worthiness, since it was never enabled for us.  Removing it as False would mean no change for edX.org, but I don't know about the external community, and they are the types of things that can be discussed in the DEPR ticket. Thank you!

robrap reviewed (approved):
Feel free to squash.  Thanks.  If you could help me more quickly validate the recommendations (as noted in my comment), that would be great.  :)

nizarmah commented:
Squashed and added the information about the values that will likely be set [after deprecation](https://github.com/edx/edx-platform/pull/26175#discussion_r566459463). 

Possible merge coming 👀 ? 🤞🏼 😃

robrap commented:
@nizarmah: I wasn't sure exactly how you wanted to squash / comment, so if you don't mind, just let me know when it is ready to merge.  Thanks.

nizarmah commented:
@robrap thanks a lot for your patience and effort reviewing this pull request ❤️ 

I squashed with a somewhat long commit message, but feel free to remove any details you don't want to include 😃

robrap commented:
Thank you for putting so much effort into this PR, and for kicking off the upcoming DEPR ticket! Both the docs and the DEPR tickets will greatly help simplify things for many to come.

edx-webhook commented:
@nizarmah 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/26175&entry.1484655318=https://github.com/nizarmah&entry.752974735=2021-01-26+16:59&entry.1917517419=2021-02-01+21:53&entry.2133058324=Yes) so we can improve your experience in the future.

nizarmah commented:
Hello @robrap @nasthagiri! I hope you had a nice weekend.

I wanted to let you know that I won't be able to create the tickets, unfortunately, because we had an incident which I ended up taking care of. I wanted to let you know that @Kelketek offered to create the DEPR tickets, instead.

Sorry about that!

robrap commented:
No worries @nizarmah. Thank you for all your work on this, and we appreciate you finding a replacement for creating the DEPR as well.

nizarmah commented:
@robrap @nasthagiri It's our pleasure to be helping out! The deprecation tickets have been posted, alongside a [forum thread](https://discuss.openedx.org/t/deprecation-removal-several-block-structure-caching-settings-depr-143-depr-144-depr-145-depr-146/4288?u=nizar) for any discussion. The deprecation tickets are also linked there. 

Hope you have a wonderful day!

robrap commented:
Great news @nizarmah. Thank you. DEPR for outdated toggles is one of the goals of documenting.

regisb commented:
TNL-5041, which is referenced in these docs, is not accessible by non-edX people: https://openedx.atlassian.net/browse/TNL-5041
Could someone eventually fix the permissions?

natabene commented:
@regisb Sorry, I don't think permissions can be changed easily, this ticket is in the archive now. @nasthagiri Do you think this ticket or entire archive can be made accessible to public? Any reason to keep it restricted?

nasthagiri commented:
A few things:

1. We cannot have feature toggle annotations link to edx-private JIRA tickets. That information is useless for the majority of our consumers. @robrap Can we update the guidelines to be explicit about this?
2. Ideally, our codebase doesn't link to JIRA tickets at all. As we can see here, the lifetime/archival of our JIRA tickets are not guaranteed. However, this may be a long-term change that will take some time for us to mature into. cc: @nedbat on why not JIRA for the long-term.
3. In the meantime, what is the best location for us to store the contents of the non-archived JIRA ticket? Making the JIRA ticket public doesn't seem to be the correct long-term answer. My proposal is to create a backdated ADR for this information and put the ADR in the `block_structure` folder.  @robrap (or others) what do you think?

robrap commented:
>     1. We cannot have feature toggle annotations link to edx-private JIRA tickets. That information is useless for the majority of our consumers. @robrap Can we update the guidelines to be explicit about this?

I added this to my list of toggle updates needed. I just need some time to organize and/or implement.
 
>     3. In the meantime, what is the best location for us to store the contents of the non-archived JIRA ticket? Making the JIRA ticket public doesn't seem to be the correct long-term answer. My proposal is to create a backdated ADR for this information and put the ADR in the `block_structure` folder.  @robrap (or others) what do you think?

ADRs are certainly a great option if someone is willing and able to write it.  In some cases, we point to a PR, and we've copied relevant information to the PR description.  This may be less ideal, but has been more expedient.

nasthagiri commented:
> 1. We cannot have feature toggle annotations link to edx-private JIRA tickets.

Is this something we can lint for, you think? @regisb

nasthagiri commented:
> 3. In some cases, we point to a PR, and we've copied relevant information to the PR description.

That's a great idea.  @nizarmah Would you be able to dig up the PR and update the annotations?  I can take the action to move over the content from the private JIRA ticket, once you've found the PR.  If there's no appropriate PR link for this, we can create an ADR.

robrap commented:
> > 1. We cannot have feature toggle annotations link to edx-private JIRA tickets.
> 
> Is this something we can lint for, you think? @regisb

@nasthagiri: I had added doc updates and linting to my list for this.  We could certainly lint for links to our JIRA instance, or short annotations with no link that look like a JIRA ticket. Note: I also already have a note for renaming this annotation, since the name is asking for a ticket.  But, we'll want to know if DEPR tickets should be an exception, or if we also want alternatives there.

nasthagiri commented:
> DEPR tickets

There is a 1:1 relationship now with DEPR tickets and Discourse tickets. We can favor linking to the Discourse ticket over the DEPR ticket.  (Over time, the DEPR board can become more of a task-management board rather than a discussion board.)

I'll ask @dianakhuang what she thinks about updating the DEPR OEP to have folks discuss the DEPR in Discourse, but use the DEPR board for managing its status.

regisb commented:
> Is this something we can lint for, you think?

@nasthagiri Yes, this is pretty easy. However, the current recommendation is actually to link to Jira ticket, in the `.. toggle_tickets` field: https://edx-toggles.readthedocs.io/en/latest/how_to/documenting_new_feature_toggles.html Should we amend that?

robrap commented:
> @nasthagiri Yes, this is pretty easy. However, the current recommendation is actually to link to Jira ticket, in the `.. toggle_tickets` field: https://edx-toggles.readthedocs.io/en/latest/how_to/documenting_new_feature_toggles.html Should we amend that?

Generally, yes, we should update the doc. We will likely want to rename (expand/contract) the annotation name, but that can come next. A quick doc update here is fine, but otherwise, let’s prioritize linting against other toggles work.

nizarmah commented:
@nasthagiri sure. There's a bunch of PRs to link to though, not a single specific one. From the toggles that have multiple pull requests, would you mind choosing which ones you'd prefer be included in the annotation for each toggle?

* [`BLOCK_STRUCTURES_SETTINGS`](https://github.com/edx/edx-platform/pull/13388)
* `BLOCK_STRUCTURES_SETTINGS['PRUNING_ACTIVE']`
  * [`PRUNE_OLD_VERSIONS` implementation](https://github.com/edx/edx-platform/pull/14571)
  * [`PRUNING_ACTIVE` tweak](https://github.com/edx/edx-platform/pull/17760)
* [`BLOCK_STRUCTURES_SETTINGS['COURSE_PUBLISH_TASK_DELAY']`](https://github.com/edx/edx-platform/pull/13388)
* `block_structure.invalidate_cache_on_publish`
  * [Initial implementation](https://github.com/edx/edx-platform/pull/14358)
  * [Conversion to waffle switch](https://github.com/edx/edx-platform/pull/14770)
* `block_structure.storage_backing_for_cache`
  * [Initial implementation](https://github.com/edx/edx-platform/pull/14512)
  * [Conversion to waffle switch](https://github.com/edx/edx-platform/pull/14770)
* `block_structure.raise_error_when_not_found`
  * [Initial implementation](https://github.com/edx/edx-platform/pull/14512)
  * [Conversion to waffle switch](https://github.com/edx/edx-platform/pull/14770)
* [`BLOCK_STRUCTURES_SETTINGS['DIRECTORY_PREFIX']`](https://github.com/edx/edx-platform/pull/14571)
* [`BLOCK_STRUCTURES_SETTINGS['STORAGE_CLASS']`](https://github.com/edx/edx-platform/pull/14571)
* [`BLOCK_STRUCTURES_SETTINGS['STORAGE_KWARGS']`](https://github.com/edx/edx-platform/pull/14571)

robrap commented:
> @nasthagiri sure. There's a bunch of PRs to link to though, not a single specific one. From the toggles that have multiple pull requests, would you mind choosing which ones you'd prefer be included in the annotation for each toggle?

1. Feel free to add anything that would be helpful to “future you” or another dev.
2. I think we’ve included multiple URLs separated by a comma in some cases.

nizarmah commented:
@robrap if you have time, I'd love to have your review on [the pull request](https://github.com/edx/edx-platform/pull/27059#issue-595392727). Let me know please 😃 cc @nasthagiri
